### ISSUE TITLE
250825 Computer Vision Basic_안희원

### ISSUE THREAD

**Issue by notone0818** [AIBootcamp14/wiki-rider#36]
## 📅 학습 날짜  
*2025-08-25*

## 📌 키워드 및 요약 정리
_[kyeword]: computer vision, backbone ,decoder_

- **Conputer Vision**은 이미지, 비디오, 라디오, 텍스트 등의 데이터를 처리하는 기술이다.
- **Backbone**은 이미지 처리에서 중요한 특징을 추출하는 모델이다.
- **Decoder**는 추출된 특징을 이미지로 변환하는 모델이다.
- **Encoder**는 이미지 처리에서 중요한 특징을 추출하는 모델이다.

### 1. What is Computer Vision?

Computer Vision은 인간의 시각을 흉내 내는 컴퓨터 프로그램이다. 실제 사례를 들어보자면 (1) Google Lens, (2) 얼굴 랜드마크 검출, (3) Image Captioning
, (4) Teachable Machine 등이 있다.

Computer Vision은 현재 **robot**과 협업하는 연구가 활발히 진행되고 있다. Computer Vision이 eye를 담당하며, robot은 hand를 담당하므로, *eye-hand cooridination*이라고도 불린다. 이 뿐만 아니라, **NLP** 분야와의 협업도 활발히 연구되고 있다. 

Computer Vision와 관련한 필자의 프로젝트 하나의 예시를 들어보자면, **Anomaly Detection CCTV**가 있다. Edge Computing에서는 실시간으로 이미지를 필터링, 가공하여 Server로 보낸 후, Server에서 이미지를 분석하여 이상이 있는지 확인한다. 물론 모든 처리를 엣지에서 수행할 수도 있지만, 필자는 서버와 엣지 컴퓨팅의 역할을 분리하여 설계하였다. 

위와 같은 예시에서 눈치챌 수 있듯, **Low-Latency**가 매우 중요하다. Google Lens나 Papago의 실시간 번역 기능에서도 지연 시간을 최소화하는 것이 핵심이다. 따라서 단순히 정확도(Accuracy)만 고려하는 것이 아니라, **경량화**와 **최적화**를 통해 주어진 자원 내에서 Latency와 Cost를 균형 있게 관리하는 설계가 필요하다.

### 2. 고전 컴퓨터 비전

고전 컴퓨터 비전에서는 딥러닝 모델 없이 데이터를 가공할 때 활용한다. 즉, 딥러닝으로 해결하기 어려운 문제에 활용이 된다. 대표적인 기법으로는 (1) Morphological Operation, (2) Edge & Contour Detection 등이 있다.

**Morphological Operation**은 이미지의 구조를 변경하는 연산으로, 주로 이미지의 모양을 변경하거나 이미지의 구조를 분석하는 데 사용된다. (1) Erosion [물체 경계 침식], (2) Dilation [물체 경계 확장], (3) Opening [침식 후 확장], (4) Closing [확장 후 침식]이 있다.

**Edge & Contour Detection**은 이미지의 경계를 찾는 연산으로, 주로 이미지의 구조를 분석하는 데 사용된다. Edge Detection에서 유명한 방법론에는 Canny Edge Dectector가 있다.

```python
# Canny Edge Detection
import cv2
import matplotlib.pyplot as plt

# 이미지 읽기 (그레이스케일)
img = cv2.imread("image.jpg", cv2.IMREAD_GRAYSCALE)

# Canny Edge Detection
edges = cv2.Canny(img, threshold1=100, threshold2=200)

# 결과 출력
plt.figure(figsize=(10,5))
plt.subplot(1,2,1)
plt.imshow(img, cmap='gray')
plt.title("Original Image")
plt.axis("off")

plt.subplot(1,2,2)
plt.imshow(edges, cmap='gray')
plt.title("Canny Edges")
plt.axis("off")

plt.show()
```

### 3. Backbone & Decoder
Computer Vision Model은 일반적으로 **Backbone**과 **Decoder** 구조로 구성된다. 
- **Backbone** -> 입력 이미지에서 유의미한 특징(feature)을 추출하고 압축하는 역할 수행. ResNet, VGG, MobileNet과 같은 사전 학습(pre-trained) 모델이 자주 사용된다. 

- **Decoder** -> 추출된 특징은 비전 태스크에 따라 Decoder를 통해 처리된다. 예를 들어, Image Classification에서는 Fully Connected Layer와 Softmax를 통해 클래스 확률을 출력하고, Object Detection에서는 Head가 특징 맵을 기반으로 바운딩 박스와 클래스 예측을 수행하며, Semantic Segmentation에서는 Encoder-Decoder 구조를 활용하여 각 픽셀 단위로 클래스 확률을 예측한다. 

- **Encoder** -> Encoder를 추가하여 특징과 이미지 패치 간 관계를 학습하기도 한다. 이러한 구조 덕분에 동일한 Backbone을 다양한 비전 태스크에 재사용할 수 있으며, Decoder는 태스크 특성에 맞춰 결과를 생성하도록 설계된다.

## 📚 참고 자료
https://www.ted.com/talks/aicha_evans_your_self_driving_robotaxi_is_almost_here

-> Amazon의 자회사 Zoox에서 부르면 달려오는 Robotaxi 서비스를 소개하는 2021년 Ted 영상입니다. 2024년 미국 방문하였을 때, 실제로 Robot이 음식 배달을 하고 있는 모습을 보았습니다. 한국은 지리적 특성상 이러한 서비스 도입이 쉽지 않아 보이지만, 언젠가는 가능하지 않을까 하는 기대도 듭니다.

https://docs.opencv.org/4.x/da/d22/tutorial_py_canny.html

-> OpenCV의 Canny Edge Detection을 소개하는 문서

## ❓ 어려웠던 부분  
- 실제 Computer Vision project를 하게 된다면 Camera 혹은 IoT나 임베디드 장치와는 떼어놓을 수 없을 거라 생각한다. 이때, CS 지식 그리고 C/C++의 지식은 매우 중요하다는 것을 깨달았다. 어디까지 배워야하고 어디까지는 필요 없는지 잘 모르겠다. 

## 💬 질문 또는 토론 유도
- `fork`와 `spawn` 방식은 프로세스 생성 구조에서 차이가 있는데, 이러한 구조적 차이가 실제 성능에 어떤 영향을 주는가?

- `WSL2`와 `Ubuntu` 환경을 비교했을 때, 어떤 상황에서는 WSL2가 충분히 효율적일 수 있지만 또 어떤 상황에서는 한계가 드러날까요? (윈도우 유저라면 리눅스 환경을 구축하기 위해 주로 WSL2나 VirtualBox를 선택할 텐데, 필자의 경우 WSL2를 사용하여 환경을 구성했습니다.)

- 나아가, CPU 코어 수나 메모리 대역폭 같은 컴퓨터 사양이 이 두 방식의 효율 차이에 얼마나 영향을 주는가?

(참고로 저는 2~3년 전에 로컬 WSL을 통해 리눅스 환경에 접속해서 사용했는데 [운영체제 : window], 그때 Windsurf만 유독 연결이 안 되어서 한참 애먹었던 기억이 있습니다. 그 경험 이후로는 VSCode는 괜찮지만, 이른바 AI 기반 코드 에디터류에 대해서는 신뢰가 많이 떨어지게 되었어요.. 하핳)

**Comment by LeeHyunJong** [comment]
안녕하세요. 개념을 잘 정리해주셨고, 프로젝트 경험과 관련한 인사이트를 흥미롭게 읽었습니다!

### 어려웠던 부분

> 실제 Computer Vision project를 하게 된다면 Camera 혹은 IoT나 임베디드 장치와는 떼어놓을 수 없을 거라 생각한다. 이때, CS 지식 그리고 C/C++의 지식은 매우 중요하다는 것을 깨달았다. 어디까지 배워야하고 어디까지는 필요 없는지 잘 모르겠다.

CV를 이용해 실제 제품을 개발한다고 하면, 말씀해주신 것처럼 C/C++ 활용뿐 아니라 운영체제와 네트워크에 대한 기초 지식이 요구됩니다. 예를 들면, 운영체제에서는 쓰레드, 메모리 관리, 동기화 개념이, 네트워크에서는 주요 프로토콜, TCP/UDP 개념이 중요할 것입니다.
"어디까지 배워야 하는가?"라는 질문에는 정답이 없다고 생각하지만, 운영체제와 네트워크의 기본기(대학 강의 수준?)를 쌓아두고, 실제 개발 과정에서 요구되는 스펙에 맞춰 심화 학습을 이어가는 것이 중요하다고 생각합니다.

### 질문

CV basics 내용이 아니라서 정확하지 않을 수 있다는 점을 먼저 말씀드립니다...

> `fork`와 `spawn` 방식은 프로세스 생성 구조에서 차이가 있는데, 이러한 구조적 차이가 실제 성능에 어떤 영향을 주는가?

제가 알기로는 fork는 부모 프로세스를 복사해서 가져오기 때문에 작은 작업을 빈번하게 할 때 유리하고 spawn은 부모 프로세스 없이 바로 프로그램을 실행하므로 새로운 프로그램을 실행할 때 유리합니다.

> `WSL2`와 `Ubuntu` 환경을 비교했을 때, 어떤 상황에서는 WSL2가 충분히 효율적일 수 있지만 또 어떤 상황에서는 한계가 드러날까요? (윈도우 유저라면 리눅스 환경을 구축하기 위해 주로 WSL2나 VirtualBox를 선택할 텐데, 필자의 경우 WSL2를 사용하여 환경을 구성했습니다.)

WSL을 사용하면서 작업 환경을 구성하는 데에 있어서 이슈가 없었습니다. 직접 Ubuntu를 사용하는 것과 비교해서는 시스템 수준에서 작업을 한다면 오버헤드가 있을 것 같습니다.

> 나아가, CPU 코어 수나 메모리 대역폭 같은 컴퓨터 사양이 이 두 방식의 효율 차이에 얼마나 영향을 주는가?

WSL은 윈도우 환경 위에 가상의 환경이므로 Ubuntu가 더욱 효율적일 것이고, 컴퓨터 사양이 좋을수록 spawn 방식으로 새로운 프로그램을 실행하는 것이 효율적일 것 같습니다.

**Comment by jphan32** [comment]
## 수정사항
- kyeword : 단순 오타... **keyword**, Conputer Vision : **Computer Vision**
- Computer Vision은 이미지, 비디오, _~~라디오, 텍스트~~_ 등의 데이터를 처리하는 기술이다. : 라디오나 텍스트는 각각 음성 처리, 자연어 처리 영역 또는 최근 멀티모달 영역에서 다룹니다. 순수 Computer vision은 이미지나 비디오와 같은 시각 정보를 다룹니다.
- Decoder는 “추출된 특징을 이미지로 변환”하는 모듈만을 뜻하지 않습니다. 분류에서는 전역 풀링과 FC 레이어, 검출에서는 박스/클래스 헤드가 디코더 역할을 합니다. 태스크별 예시를 곁들이면 혼동이 줄어듭니다.

## 추가 설명
- 대부분의 현대 비전 모델은 Backbone에서 feature map을 만들고, Decoder/Head에서 태스크별 출력을 생성합니다. ResNet+FPN을 공통으로 두고 분류, 검출, 세그멘테이션 헤드를 바꿔 끼우는 구조를 꼭 체크해 보세요.
- Canny Edge Detector는 Gaussian 블러 → 그래디언트 계산 → non-maximum suppression → hysteresis threshold 네 단계를 거칩니다. 단계별 의도를 이해하면 왜 고전 기법이 여전히 쓰이는지 감이 옵니다.

## 질문에 대한 답변
- `fork`는 부모 프로세스를 그대로 복사해 실행을 이어가므로 초기화가 빠르고, POSIX 기반 Linux에서 데이터로더가 기본으로 선택합니다. Windows나 일부 환경에서는 `spawn`이 기본이라 초기 로딩 비용이 더 큰 점을 기억해 주세요.
- WSL2는 가상화 계층이 있어 GPU 패스스루, 파일 I/O에서 약간의 오버헤드가 생깁니다. CUDA 12 이후 학습은 가능하지만, USB·네트워크 장치를 직접 다룰 땐 네이티브 Ubuntu가 더 편합니다.
- CPU 코어 수와 메모리 대역폭은 데이터로더 워커 수와 직결됩니다. 코어 수-1 정도로 `num_workers`를 맞추고, 병목이 생기면 프로파일링으로 확인해 보세요.

## 학습 팁
- 동일한 프로젝트를 WSL2와 네이티브 Ubuntu 모두에 세팅해 학습 속도, I/O 시간을 비교표로 남겨 보세요.
- Edge/Server 분산 설계를 연습할 때는 단계마다 지연 시간을 기록해 병목을 찾는 습관을 들이면 좋습니다.
- 운영체제 지식은 필요 지점부터 토막 학습해도 충분합니다. 우선순위로 프로세스/스레드, 파일 시스템, 네트워크 소켓을 정리해 보세요.
